onnxruntime/core/framework/prepacked_weights_container.h

```diff
@@ -66,4 +73,163 @@ class PrepackedWeightsContainer final {
+      return parent_;
+    }
+
+    Subgraph& GetOrCreateSubgraph(const Graph& graph) {
+      auto result = subgraph_prepacks_.emplace(&graph, nullptr);
+      if (result.second) {
+        result.first->second = std::make_unique<Subgraph>(this, key_to_blobs_, save_mode_on_);
+      }
+      return *result.first->second;
+    }
+
+    const Subgraph* GetSubgraph(const Graph& graph) const {
+      auto it = subgraph_prepacks_.find(&graph);
+      return it == subgraph_prepacks_.end() ? nullptr : it->second.get();
+    }
+
+    void InsertFromDisk(const std::string& key, PrePackedWeights&& packed_weight);
+
+    void WritePacked(const std::string& weight_name, const std::string& key,
+                     PrePackedWeights&& packed_weight);
+
+    const PrePackedWeights* GetPrepackedWeights(const std::string& key) const;
+
+    // The function would add or replace existing entry with references to it.
+    // If the entry is present, it would replace it with references to the existing entry.
+    // If the entry is not present, it would add reference to refer_if_absent
+    // If the entry is present it would return the existing entry otherwise std::nullopt
+    std::optional<PrePackedWeights> ReplaceWithReferenceIfSaving(const std::string& weight_name,
+                                                                 const std::string& key,
+                                                                 const PrePackedWeights& refer_if_absent);
+
+    bool IsSaveModeOn() const noexcept {
+      return save_mode_on_;
+    }
+
+    void SetSaveMode(bool value) noexcept {
+      save_mode_on_ = value;
+    }
+
+    // Returns iterators to key->blob pair for writing
+    const KeysPerWeight* GetBlobsForWeight(const std::string& weight_name) const {
+      auto hit = sorted_by_weight_for_writing_.find(weight_name);
+      if (hit != sorted_by_weight_for_writing_.end()) {
+        return &hit->second;
+      }
+      return nullptr;
+    }
+
+    size_t GetSubgraphNum() const noexcept {
+      return subgraph_prepacks_.size();
+    }
+
+    size_t GetNumberOfWeightsForWriting() const noexcept {
+      return sorted_by_weight_for_writing_.size();
+    }
+
+    size_t GetNumberOfKeyedBlobsForWriting() const noexcept {
+      size_t result = 0;
+      for (const auto& [_, keys] : sorted_by_weight_for_writing_) {
+        result += keys.size();
+      }
+      return result;
+    }
+
+    const WeightToPrePacksMap& GetSortedByWeightForWriting() const noexcept {
+      return sorted_by_weight_for_writing_;
+    }
+
+    // This template declaration is intended to be instantiated
+    // and used by test code to inspect private members. Otherwise,
+    // it has no body and is not intended to be used by other code.
+    template <class T>
+    void TestHarness(T&) const;
+
+   private:
+    bool save_mode_on_ = false;
+    Subgraph* parent_ = nullptr;
+    KeyToBlobMap& key_to_blobs_;
+    WeightToPrePacksMap sorted_by_weight_for_writing_;
```

Comment:
The name sorted_by_weight_for_writing_ is confusing. I have not seen the sorting logic.